web/src/app/components/cdf/cdf.component.ts

```diff
@@ -0,0 +1,222 @@
+ */
+type d3SVG = d3.Selection<SVGSVGElement, undefined, null, undefined>;
+
+@Component({
+  selector: 'app-cdf',
+  templateUrl: './cdf.component.html',
+  styleUrls: ['./cdf.component.scss']
+})
+
+export class CdfComponent implements AfterViewInit {
+  private static readonly NANO_TO_MINUTES: number = (10 ** 9) * 60;
+  private static readonly COMPLETED_BLUE: string = '#00bfa5';
+  private static readonly COMPLETED_STATE_TAG: number = 5;
+
+  @ViewChild('cdf') private cdfContainer!: ElementRef;
+  @Input() pushInfos!: step189_2020.IPushInfo[] | null;
+  @Input() currentPush!: step189_2020.IPushInfo | null;
+
+  private data: Item[] = [];
+  private graphData: Item[] = [];
+  private svg: d3SVG | undefined;
+
+  /**
+   * Calculates the duration between the completed stage and the first non-empty
+   * stage. Assigns the probability as the rank of the duration value over the
+   * total number of points. The duration and probability are defined in an
+   * interface and all points stored as an array of CdfData interfaces.
+   * 
+   * @param pushInfos Array of pushes for a single push def
+   * @return Array of Items sorted by increasing duration
+   */
+  private static populateData(pushInfos: step189_2020.IPushInfo[]): Item[] {
+    const durations: number[] = [];
+    pushInfos.forEach(pushInfo => {
+      if (!pushInfo) { return; }
+      const states = pushInfo.stateInfo;
+      if (!states) { return; }
+      const pushEndTime = states[states.length - 1].startTimeNsec;
+      if (!pushEndTime) { return; }
+      const finalState = states[states.length - 1].state;
+      if (!finalState) { return; }
+
+      if (finalState === CdfComponent.COMPLETED_STATE_TAG) {
+        // Find the start time of the first non-empty stage.
+        let firstStateStart: number | Long = -1;
+        for (const state of states) {
+          if (state.stage && state.startTimeNsec) {
+            firstStateStart = state.startTimeNsec;
+            break;
+          }
+        }
+        if (firstStateStart !== -1) {
+          const duration = (+pushEndTime - +firstStateStart) / CdfComponent.NANO_TO_MINUTES;
+          durations.push(duration);
+        }
+      }
+    });
+
+    const sortedArray: number[] = durations.sort((n1, n2) => n1 - n2);
+
+    const data: Item[] = [];
+    const durationLength = sortedArray.length;
+    for (let i = 0; i < durationLength; i++) {
+      const duration = sortedArray[i];
+      const probability = (i + 1) / durationLength;
+      data.push({
+        duration,
+        probability
+      } as Item);
+    }
+
+    return data;
+  }
+
+  /**
+   * Creates a CDF chart by plotting the duration of completed pushes against 
+   * the probability of a push taking less time than that duration.
+   * 
+   * Structure of the SVG:
+   * CdfChart:
+   *     - y-axis-left, y-axis-right, y-axis-left-label
+   *     - path:
+   *         - cdf-curve
+   * svg:
+   *     - x-axis, x-axis-label, graph-title
```

Comment:
What I had in mind is an expanded version of something like:

```
<svg>
  <g></g>
  <g class='x-axis'>
      [...]
  </g>
  <text class='x-axis-label'>[...]</text>
  <text class='graph-title'>[...]</text>
</svg>
```

Reply:
I added the base level tags but decided to leave out some of the more in-depth components, like the axes ticks